README.md
# age_calculator

A Flutter package that allows you to perform operations between dates:
- Calculate a person's age and the time remaining to their next birthday in years, months and days. 
- Calculate differences between two dates in years, months and days.
- Add years, months and days to any date.

## Getting Started

In your flutter project add the dependency:

dependencies:
  ...
  age_calculator: ^1.0.0

For help getting started with Flutter, view the online documentation.

### Example

```dart
import 'package:age_calculator/age_calculator.dart';

void main() {
  DateTime birthday = DateTime(1997, 3, 5);

  DateDuration duration;

  // Find out your age as of today's date 2021-03-08
  duration = AgeCalculator.age(birthday);
  print('Your age is $duration'); // Your age is Years: 24, Months: 0, Days: 3

  //Find out your age on any given date
  duration = AgeCalculator.age(birthday, today: DateTime(2030, 5, 1));
  print('Your age is $duration'); // Your age is Years: 33, Months: 1, Days: 26

  // Find out when your next birthday will be at 2021-03-08
  duration = AgeCalculator.timeToNextBirthday(birthday);
  print('You next birthday will be in $duration');
  // You next birthday will be in Years: 0, Months: 11, Days: 25

  // Find out when your next birthday will be on any given date
  duration = AgeCalculator.timeToNextBirthday(birthday,
      fromDate: DateTime(2021, 3, 2));
  print('You next birthday will be in $duration');
  // You next birthday will be in Years: 0, Months: 0, Days: 3

  // Find out the difference between two dates
  duration = AgeCalculator.dateDifference(
    fromDate: DateTime(2021, 1, 2),
    toDate: DateTime(2025, 5, 2),
  );
  print('The difference is $duration');
  // You next birthday will be in Years: 4, Months: 4, Days: 0

  // Add time to any date
  DateTime date = AgeCalculator.add(
      date: DateTime(2021, 1, 2),
      duration: DateDuration(years: 5, months: 2, days: 1));
  print(date);
  // 2026-03-03 00:00:00.000
}
```


CHANGELOG.md
## [1.0.0] - 08/03/2021

Migration to null safety from package [age](https://pub.dev/packages/age) and added test coverage and new apis



test/age_calculator_test.dart
import 'package:age_calculator/age_calculator.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  var birthdate;
  var today;

  setUp(() {
    birthdate = DateTime(1997, 3, 5);
    today = DateTime(2021, 3, 8);
  });

  group('Find out', () {
    test('your age', () {
      expect(AgeCalculator.age(birthdate, today: today).toString(),
          DateDuration(years: 24, months: 0, days: 3).toString());
    });
    test('your next birthday', () {
      expect(
          AgeCalculator.timeToNextBirthday(birthdate, fromDate: today)
              .toString(),
          DateDuration(years: 0, months: 11, days: 25).toString());
    });

    test('difference between two dates', () {
      expect(
          AgeCalculator.dateDifference(toDate: today, fromDate: birthdate)
              .toString(),
          DateDuration(years: 24, months: 0, days: 3).toString());
    });
  });

  group('Add', () {
    test('days', () {
      expect(
          AgeCalculator.add(
              date: DateTime(2021, 3, 5), duration: DateDuration(days: 36)),
          DateTime(2021, 4, 10));
    });
    test('months', () {
      expect(
          AgeCalculator.add(
              date: DateTime(2021, 3, 5), duration: DateDuration(months: 20)),
          DateTime(2022, 11, 5));
    });

    test('years', () {
      expect(
          AgeCalculator.add(
              date: DateTime(2021, 3, 5), duration: DateDuration(years: 10)),
          DateTime(2031, 3, 5));
    });

    test('days, months and years', () {
      expect(
          AgeCalculator.add(
              date: DateTime(2021, 3, 5),
              duration: DateDuration(
                years: 10,
                months: 3,
                days: 5,
              )),
          DateTime(2031, 6, 10));
    });
  });
}


lib/age_calculator.dart
library age_calculator;

class DateDuration {
  int days;
  int months;
  int years;

  DateDuration({this.days = 0, this.months = 0, this.years = 0});

  String toString() {
    return 'Years: $years, Months: $months, Days: $days';
  }
}

/// Age Class
class AgeCalculator {
  /// _daysInMonth cost contains days per months; daysInMonth method to be used instead.
  static const List<int> _daysInMonth = [
    31, // Jan
    28, // Feb, it varies from 28 to 29
    31,
    30,
    31,
    30,
    31,
    31,
    30,
    31,
    30,
    31 // Dec
  ];

  /// isLeapYear method
  static bool isLeapYear(int year) =>
      (year % 4 == 0) && ((year % 100 != 0) || (year % 400 == 0));

  /// daysInMonth method
  static int daysInMonth(int year, int month) =>
      (month == DateTime.february && isLeapYear(year))
          ? 29
          : _daysInMonth[month - 1];

  /// dateDifference method
  static DateDuration dateDifference({
    required DateTime fromDate,
    required DateTime toDate,
  }) {
    // Check if toDate to be included in the calculation
    DateTime endDate = toDate;

    int years = endDate.year - fromDate.year;
    int months = 0;
    int days = 0;

    if (fromDate.month > endDate.month) {
      years--;
      months = (DateTime.monthsPerYear + endDate.month - fromDate.month);

      if (fromDate.day > endDate.day) {
        months--;
        days = daysInMonth(fromDate.year + years,
                ((fromDate.month + months - 1) % DateTime.monthsPerYear) + 1) +
            endDate.day -
            fromDate.day;
      } else {
        days = endDate.day - fromDate.day;
      }
    } else if (endDate.month == fromDate.month) {
      if (fromDate.day > endDate.day) {
        years--;
        months = DateTime.monthsPerYear - 1;
        days = daysInMonth(fromDate.year + years,
                ((fromDate.month + months - 1) % DateTime.monthsPerYear) + 1) +
            endDate.day -
            fromDate.day;
      } else {
        days = endDate.day - fromDate.day;
      }
    } else {
      months = (endDate.month - fromDate.month);

      if (fromDate.day > endDate.day) {
        months--;
        days = daysInMonth(fromDate.year + years, (fromDate.month + months)) +
            endDate.day -
            fromDate.day;
      } else {
        days = endDate.day - fromDate.day;
      }
    }

    return DateDuration(days: days, months: months, years: years);
  }

  /// add method
  static DateTime add(
      {required DateTime date, required DateDuration duration}) {
    int years = date.year + duration.years;
    years += ((date.month + duration.months) ~/ DateTime.monthsPerYear);
    int months = ((date.month + duration.months) % DateTime.monthsPerYear);

    int days = date.day + duration.days - 1;

    return DateTime(years, months, 1).add(Duration(days: days));
  }

  static DateDuration age(DateTime birthdate, {DateTime? today}) {
    return dateDifference(fromDate: birthdate, toDate: today ?? DateTime.now());
  }

  static DateDuration timeToNextBirthday(DateTime birthdate,
      {DateTime? fromDate}) {
    DateTime endDate = fromDate ?? DateTime.now();
    DateTime tempDate = DateTime(endDate.year, birthdate.month, birthdate.day);
    DateTime nextBirthdayDate = tempDate.isBefore(endDate)
        ? AgeCalculator.add(date: tempDate, duration: DateDuration(years: 1))
        : tempDate;
    return dateDifference(fromDate: endDate, toDate: nextBirthdayDate);
  }
}


example/lib/main.dart
import 'package:age_calculator/age_calculator.dart';

void main() {
  DateTime birthday = DateTime(1997, 3, 5);

  DateDuration duration;

  // Find out your age as of today's date 2021-03-08
  duration = AgeCalculator.age(birthday);
  print('Your age is $duration'); // Your age is Years: 24, Months: 0, Days: 3

  //Find out your age on any given date
  duration = AgeCalculator.age(birthday, today: DateTime(2030, 5, 1));
  print('Your age is $duration'); // Your age is Years: 33, Months: 1, Days: 26

  // Find out when your next birthday will be at 2021-03-08
  duration = AgeCalculator.timeToNextBirthday(birthday);
  print('You next birthday will be in $duration');
  // You next birthday will be in Years: 0, Months: 11, Days: 25

  // Find out when your next birthday will be on any given date
  duration = AgeCalculator.timeToNextBirthday(birthday,
      fromDate: DateTime(2021, 3, 2));
  print('You next birthday will be in $duration');
  // You next birthday will be in Years: 0, Months: 0, Days: 3

  // Find out the difference between two dates
  duration = AgeCalculator.dateDifference(
    fromDate: DateTime(2021, 1, 2),
    toDate: DateTime(2025, 5, 2),
  );
  print('The difference is $duration');
  // You next birthday will be in Years: 4, Months: 4, Days: 0

  // Add time to any date
  DateTime date = AgeCalculator.add(
      date: DateTime(2021, 1, 2),
      duration: DateDuration(years: 5, months: 2, days: 1));
  print(date);
  // 2026-03-03 00:00:00.000
}
